web/auth/magic_link.go

```diff
@@ -70,3 +75,81 @@ func loginWithMagicLink(c echo.Context) error {
 	}
 	return c.Redirect(http.StatusSeeOther, redirect.String())
 }
+
+type magicLinkFlagshipParameters struct {
+	ClientID          string `json:"client_id"`
+	ClientSecret      string `json:"client_secret"`
+	Code              string `json:"code"`
+	TwoFactorPasscode string `json:"two_factor_passcode"`
+	TwoFactorToken    string `json:"two_factor_token"`
+}
+
+func magicLinkFlagship(c echo.Context) error {
+	inst := middlewares.GetInstance(c)
+
+	var args magicLinkFlagshipParameters
+	if err := c.Bind(&args); err != nil {
+		return jsonapi.Errorf(http.StatusBadRequest, "%s", err)
+	}
+
+	if err := lifecycle.CheckMagicLink(inst, args.Code); err != nil {
+		err := config.GetRateLimiter().CheckRateLimit(inst, limits.MagicLinkType)
+		if limits.IsLimitReachedOrExceeded(err) {
+			return echo.NewHTTPError(http.StatusTooManyRequests, "Too many requests")
+		}
+		return c.JSON(http.StatusUnauthorized, echo.Map{
+			"error": inst.Translate(CredentialsErrorKey),
+		})
+	}
+
+	if inst.HasAuthMode(instance.TwoFactorMail) {
+		// TODO 2FA
+		return c.JSON(http.StatusUnauthorized, echo.Map{
+			"error": inst.Translate(CredentialsErrorKey),
+		})
+	}
+
+	client, err := oauth.FindClient(inst, args.ClientID)
+	if err != nil {
+		if couchErr, isCouchErr := couchdb.IsCouchError(err); isCouchErr && couchErr.StatusCode >= 500 {
+			return err
+		}
+		return c.JSON(http.StatusBadRequest, echo.Map{
+			"error": "the client must be registered",
+		})
+	}
+	if subtle.ConstantTimeCompare([]byte(args.ClientSecret), []byte(client.ClientSecret)) == 0 {
+		return c.JSON(http.StatusBadRequest, echo.Map{
+			"error": "invalid client_secret",
+		})
+	}
+
+	if !client.Flagship {
+		return ReturnSessionCode(c, http.StatusAccepted, inst)
+	}
```

Comment:
No, it can still be the flagship app that is making the request but the stack doesn't know it. It happens when the store certification has failed. In that case, we are going to a middle path, using the magic link to allow the user to manually certify that it is really the flagship app.